net/src/ipv6/prefix.rs

```diff
@@ -0,0 +1,669 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+use crate::ipv4::Contains;
+#[cfg(any(test, feature = "bolero"))]
+#[allow(unused_imports)] // re-export
+pub use contract::*;
+use ipnet::{AddrParseError, Ipv6Net};
+use prefix_trie::Prefix;
+use serde::{Deserialize, Serialize};
+use std::fmt::{Debug, Display, Formatter};
+use std::net::Ipv6Addr;
+use std::str::FromStr;
+use tracing::{debug, warn};
+
+/// An `Ipv6Addr` with a mask describing a network in CIDR notation.
+///
+/// Note that unlike [`Ipv6Net`] from the `ipnet` crate, this type ensures that only network bits
+/// are set in the address.
+#[derive(
+    Default, Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
+)]
+#[repr(transparent)]
+#[serde(transparent)]
+pub struct Ipv6Prefix(Ipv6Net);
+
+/// A checked type describing the values 0 to 128, which constitute all legal prefix lengths for
+/// Ipv6 addresses.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[repr(transparent)]
+#[serde(try_from = "u8", into = "u8")]
+pub struct Ipv6PrefixLen(u8);
+
+/// An error indicating that an invalid prefix length was provided.
+#[derive(Debug, thiserror::Error)]
+#[repr(transparent)]
+pub enum InvalidIpv6PrefixLength {
+    /// The provided prefix is too long to form a legal [`Ipv6PrefixLen`]
+    #[error("invalid prefix length {0} is invalid, max is {MAX}", MAX = Ipv6PrefixLen::MAX_LEN)]
+    TooLong(u8),
+}
+
+/// An error indicating that an invalid network was provided.
+#[derive(Debug, thiserror::Error)]
+pub enum InvalidIpv6Network {
+    /// The provided network description contains set non-network bits
+    #[error("Address {0}/{1} contains non network bits")]
+    AddressContainsNonNetworkBits(Ipv6Addr, Ipv6PrefixLen),
+    /// The provided prefix length is invalid
+    #[error(transparent)]
+    InvalidPrefix(InvalidIpv6PrefixLength),
+}
+
+/// An error indicating that an invalid network was provided.
+#[derive(Debug, thiserror::Error)]
+pub enum Ipv6PrefixParseError {
+    /// failure to interpret string as an ip and a prefix length
+    #[error(transparent)]
+    AddrParseError(AddrParseError),
+    /// invalid ip or prefix length
+    #[error(transparent)]
+    InvalidIpv6Network(InvalidIpv6Network),
+}
+
+impl Ipv6PrefixLen {
```

Comment:
Method suggestion: `pub fn to_prefix_size(&self) -> PrefixSize` (for a follow-up :slightly_smiling_face: or I may end up adding it when rebasing my PR).

Reply:
Yup.  Will leave that for another PR tho